Convert Chapel strings to Python strings, support embedded nulls

dlongnecke-cray opened the pull request:
This PR adjusts Python interop code to convert Chapel strings returned from exported routines to Python strings. Previously, they were returned as Python bytes, and only Python bytes could be passed to exported Chapel routines with string parameters.

The Chapel string now maps cleanly to the Python string in Python interop. Additionally, this PR also supports passing strings containing null bytes between Chapel and Python. Previously, we used strlen to get the length of a Chapel string (as it was exported to C as a `char*`), but now both strings and bytes are shuffled around as a `chpl_byte_buffer`. 

---

Testing:

- [x] `ALL` on `linux64` when `CHPL_COMM=none` (no `cython`/`numpy`)
- [x] `interop` on `linux64` when `CHPL_COMM=none`
- [x] `interop` on `linux64` when `CHPL_COMM=gasnet`
- [x] `interop` on `darwin` when `CHPL_COMM=none`

dlongnecke-cray commented:
Since I'm already halfway there, I'm just going to finish supporting embedded NULLs on this branch, rather than waiting for this branch to be merged first.

dlongnecke-cray commented:
That was easier than I expected! I think this effort is ready for review.

lydia-duncan commented on compiler/codegen/mli.cpp:
```diff
@@ -906,13 +906,13 @@ std::string MLIContext::genServersideRPC(FnSymbol* fn) {
   std::string gen;
 
-  if (t == dtStringC || t == exportTypeCharPtr) {
+  if (t == dtStringC) {
```
Do we not clean up other pointers now?

dlongnecke-cray replied:
We no longer use `c_ptr(c_char)` as a target type in interop, so that would leave `c_string` as the only pointer type with buffer-like semantics, I believe?

lydia-duncan replied:
Ah, okay, I'd missed that connection

lydia-duncan commented on compiler/codegen/symbol.cpp:
```diff
@@ -1110,18 +1112,45 @@ std::string ArgSymbol::getPythonArgTranslation() {
+    res += ":\n";
+    res += "\t\traise TypeError(\"Expected \'";
+    res += pythonType;
+    res += "\' in conversion to \'chpl_byte_buffer\', ";
```
I probably should have caught this when it was originally added, but it might be nice to give an error using the Chapel type name instead of our under-the-covers conversion type name

dlongnecke-cray replied:
That's a good idea.

lydia-duncan commented on compiler/codegen/symbol.cpp:
```diff
@@ -2227,19 +2256,52 @@ GenRet FnSymbol::codegenPYXType() {
 
-      // This will free the "chpl_bytes_wrapper" buffer if required.
-      returnStmt += "\tchpl_bytes_wrapper_free(rdat)\n";
+      // Decode the bytes into a Python string.
```
Did we decide to convert c_strings into Python strings as well?  That seems inconsistent with Cython's normal behavior towards c_strings.

dlongnecke-cray replied:
The question I was asking myself is what we want Chapel to map the `c_string` to in Python interop (rather than what Cython does, as I believe I can brute force my way around that). It seems reasonable to me to map C strings to Python strings.

I'd be happy to revert if we want to map to bytes.

For instance, I didn't consider existing code that might break due to this change (again 🙄).

lydia-duncan replied:
I think it'd be good to follow the Cython behavior, since 1) c_string should go away eventually, and 2) it's really intended to be "just a C string" so decisions Cython made about C's strings should be respected

dlongnecke-cray replied:
Ok, I'll continue to convert `c_string` to bytes.

lydia-duncan commented on modules/internal/ExportWrappers.chpl:
```diff
@@ -23,65 +23,52 @@ module ExportWrappers {
+    result.size = val.size:uint(64);
+    // Assume ownership of the string/bytes record's internal buffer.
+    val.isowned = false;
+    result.size = val.size:uint(64);
```
You change the size twice here, I don't think that's necessary?

dlongnecke-cray replied:
Good catch, thanks!

lydia-duncan commented on runtime/include/python/chplrt.pxd:
```diff
@@ -27,10 +27,16 @@ cdef extern from "chpl-external-array.h":
+cdef extern from "chpl-mem.h":
+	void chpl_free(void* ptr)
+
+cdef extern from "string.h":
```
Is there not a way to import this from something Cython provides, since it is a common C function as opposed to something we provide?

dlongnecke-cray replied:
Now that I think about it, I believe I can go back to the original code for `c_string`. This was an artifact on this branch from before I mapped the Chapel string to a byte buffer.

Thanks!

lydia-duncan commented on compiler/codegen/symbol.cpp:
```diff
@@ -2282,27 +2272,6 @@ GenRet FnSymbol::codegenPYXType() {
-      returnStmt += "\tcdef Py_ssize_t rsize = strlen(rdata)\n";
-
-      // Create a new Python bytes that is a copy of the Chapel string.
-      returnStmt += "\tret = PyBytes_FromStringAndSize(rdata, rsize)\n";
```
Does this mean we no longer need to include it in the generated cython file?

dlongnecke-cray replied:
No, not at this point, but it does seem like something we're going to have to include eventually, no? (I'm not losing sleep over it.)

lydia-duncan replied:
Okay.  Not the end of the world to include it, then!

dlongnecke-cray commented:
@lydia-duncan 

I've addressed your review feedback. Here are some of the things I've done:

- Reverted changes to `c_string`. The `c_string` type maps as before, to Python `bytes.
- Adopted a more robust approach towards passing `chpl_byte_buffer` and other interop structs by value. Interop structs are marked with the `FLAG_EXPORT_WRAPPER` to indicate they should not be passed by reference. This is a very clean way of doing things.
- Added C "constructor functions" for `chpl_byte_buffer` to make things a _bit_ more convenient for the C programmer.
- Cleaned up the conversion calls in `modules/internal/ExportWrappers.chpl` and added another argument which determines if a copy should be made. Copies are currently made for parameters with the `_in_` intent, but there is room to disable them in the future (i.e., to optimize for multilocale interop).
